<a href="https://colab.research.google.com/github/SvetlanaTsim/libs_for_ds_2/blob/main/lesson_01/sv_tsim_lib_for_ds_2_hw_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Вебинар 1. Pandas


## 1 Практика


Данные для ДЗ можете скачать с google диска: https://drive.google.com/file/d/1MpWBFIbqu4mbiD0BBKYX6YhS-f4mN3Z_



<ol>
	<li>Загрузить, посмотреть, определить количество строк и объединить 3 датасета: marketing_campaign.csv, users.csv и subscribers.csv.</li>
	<li>Определить типы и&nbsp;статистики колонок</li>
	<li>Определить эффективность маркетинговых каналов по привлечению платящих игроков</li>
	<li>Определить количество игроков в каждой возрастной группе</li>
	<li>Определить самую раннюю дату подписки на сервис</li>
	<li>Определить портрет аудитории удержанных подписчиков (по возрасту и языку)</li>
</ol>

In [53]:
import numpy as np
import pandas as pd

from scipy.stats import mode

In [54]:
!wget 'https://drive.google.com/uc?id=1MpWBFIbqu4mbiD0BBKYX6YhS-f4mN3Z_' -O data.zip

--2022-12-02 14:55:37--  https://drive.google.com/uc?id=1MpWBFIbqu4mbiD0BBKYX6YhS-f4mN3Z_
Resolving drive.google.com (drive.google.com)... 108.177.125.138, 108.177.125.102, 108.177.125.139, ...
Connecting to drive.google.com (drive.google.com)|108.177.125.138|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://doc-04-c0-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/5orbabbun3dk3gok78gjrmhjrlq9n9i9/1669992900000/14904333240138417226/*/1MpWBFIbqu4mbiD0BBKYX6YhS-f4mN3Z_?uuid=766d3498-28c5-460f-9779-8882ec436817 [following]
--2022-12-02 14:55:38--  https://doc-04-c0-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/5orbabbun3dk3gok78gjrmhjrlq9n9i9/1669992900000/14904333240138417226/*/1MpWBFIbqu4mbiD0BBKYX6YhS-f4mN3Z_?uuid=766d3498-28c5-460f-9779-8882ec436817
Resolving doc-04-c0-docs.googleusercontent.com (doc-04-c0-docs.googleusercontent.com)... 108.177.97.132, 2404:6800:4008:c00::84
Connecting to do

In [55]:
!unzip data.zip

Archive:  data.zip
replace marketing description.txt? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: marketing description.txt  
replace marketing_campaign.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: marketing_campaign.csv  
replace subscribers.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: subscribers.csv         
replace users.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: users.csv               


In [56]:
#посмотрим файл с описанием
!cat 'marketing description.txt'

1) user_id: идентификатор пользователя

2) date_served: дата события

3) marketing_channel: рекламный канал

4) variant: показанный вариант рекламы

5) converted: конверсия в покупку, 1 - пользователь совершил покупку, 0 - нет

6) language_displayed: язык рекламного сообщения

7) language_preferred: предпочитаемый пользователем язык

8) age_group: возрастная группа пользователя

9) date_subscribed: дата подписки на сервис

10) date_canceled: дата отказа от подписки

11) subscribing_channel: с какого рекламного канал пришел пользователь, когда подписался на сервис

12) is_retained: удержание, 1 - пользователь продолжает пользоваться услугами сервиса, 0 - пользователь отвалился 

Источник: https://www.kaggle.com/kavitabhagwani/marketing-campaign

In [57]:
markiting_df = pd.read_csv('marketing_campaign.csv')
subscribers_df = pd.read_csv('subscribers.csv')
users_df = pd.read_csv('users.csv')

In [58]:
markiting_df.head()

user_id date_served marketing_channel          variant language_displayed  \
0   a1000      1/1/18         House Ads  personalization            English   
1   a1001      1/1/18         House Ads  personalization            English   
2   a1002      1/1/18         House Ads  personalization            English   
3   a1003      1/1/18         House Ads  personalization            English   
4   a1004      1/1/18         House Ads  personalization            English   

  converted  
0      True  
1      True  
2      True  
3      True  
4      True

In [59]:
markiting_df.sample(5)

user_id date_served marketing_channel          variant  \
630    a1630     1/15/18         House Ads  personalization   
5171   a6171      1/4/18         House Ads          control   
2650   a3650     1/21/18          Facebook          control   
4014   a5014     1/15/18              Push          control   
4505   a5505      1/2/18          Facebook          control   

     language_displayed converted  
630             English     False  
5171            English     False  
2650            English     False  
4014            English     False  
4505            English     False

In [60]:
markiting_df.describe()

user_id date_served marketing_channel  variant language_displayed  \
count    10037       10021             10022    10037              10037   
unique   10037          31                 5        2                  4   
top      a1000     1/15/18         House Ads  control            English   
freq         1         789              4733     5091               9793   

       converted  
count      10022  
unique         2  
top        False  
freq        8946

In [61]:
markiting_df.describe

<bound method NDFrame.describe of       user_id date_served marketing_channel          variant  \
0       a1000      1/1/18         House Ads  personalization   
1       a1001      1/1/18         House Ads  personalization   
2       a1002      1/1/18         House Ads  personalization   
3       a1003      1/1/18         House Ads  personalization   
4       a1004      1/1/18         House Ads  personalization   
...       ...         ...               ...              ...   
10032  a11032     1/17/18             Email          control   
10033  a11033     1/17/18             Email          control   
10034  a11034      1/5/18         Instagram          control   
10035  a11035     1/17/18             Email          control   
10036  a11036      1/6/18          Facebook          control   

      language_displayed converted  
0                English      True  
1                English      True  
2                English      True  
3                English      True  
4           

In [62]:
markiting_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10037 entries, 0 to 10036
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   user_id             10037 non-null  object
 1   date_served         10021 non-null  object
 2   marketing_channel   10022 non-null  object
 3   variant             10037 non-null  object
 4   language_displayed  10037 non-null  object
 5   converted           10022 non-null  object
dtypes: object(6)
memory usage: 470.6+ KB


In [63]:
markiting_df.shape

(10037, 6)

In [64]:
#все данные типа object
markiting_df.dtypes

user_id               object
date_served           object
marketing_channel     object
variant               object
language_displayed    object
converted             object
dtype: object

In [65]:
subscribers_df.head()

user_id subscribing_channel date_subscribed date_canceled is_retained
0   a1000           House Ads          1/1/18           NaN        True
1   a1001           House Ads          1/1/18           NaN        True
2   a1002           House Ads          1/1/18           NaN        True
3   a1003           House Ads          1/1/18           NaN        True
4   a1004           House Ads          1/1/18           NaN        True

In [66]:
subscribers_df.sample(5)

user_id subscribing_channel date_subscribed date_canceled is_retained
181    a1181           House Ads          1/9/18           NaN        True
2382   a3382                 NaN             NaN           NaN         NaN
9539  a10539                 NaN             NaN           NaN         NaN
6003   a7003                 NaN             NaN           NaN         NaN
960    a1960                Push          1/4/18       1/11/18       False

In [67]:
subscribers_df.describe()

user_id subscribing_channel date_subscribed date_canceled is_retained
count    10037                1856            1856           577        1856
unique   10037                   5              31           115           2
top      a1000           Instagram         1/16/18        4/2/18        True
freq         1                 600             163            15        1279

In [68]:
#все данные типа object
subscribers_df.dtypes

user_id                object
subscribing_channel    object
date_subscribed        object
date_canceled          object
is_retained            object
dtype: object

In [69]:
users_df.head()

user_id    age_group language_preferred
0   a1000   0-18 years            English
1   a1001  19-24 years            English
2   a1002  24-30 years            English
3   a1003  30-36 years            English
4   a1004  36-45 years            English

In [70]:
users_df.sample(5)

user_id    age_group language_preferred
8755   a9755  24-30 years            English
4765   a5765  45-55 years            English
2073   a3073  36-45 years            English
1982   a2982  24-30 years            English
9268  a10268  36-45 years            English

In [71]:
users_df.describe()

user_id    age_group language_preferred
count    10037        10037              10037
unique   10037            7                  4
top      a1000  19-24 years            English
freq         1         1682               9275

In [72]:
users_df.dtypes

user_id               object
age_group             object
language_preferred    object
dtype: object

Объединим по полю user_id (outer join, чтобы данные не потерялись)

In [73]:
full_df = pd.merge(markiting_df, subscribers_df, on='user_id', how='outer')

full_df.head()

user_id date_served marketing_channel          variant language_displayed  \
0   a1000      1/1/18         House Ads  personalization            English   
1   a1001      1/1/18         House Ads  personalization            English   
2   a1002      1/1/18         House Ads  personalization            English   
3   a1003      1/1/18         House Ads  personalization            English   
4   a1004      1/1/18         House Ads  personalization            English   

  converted subscribing_channel date_subscribed date_canceled is_retained  
0      True           House Ads          1/1/18           NaN        True  
1      True           House Ads          1/1/18           NaN        True  
2      True           House Ads          1/1/18           NaN        True  
3      True           House Ads          1/1/18           NaN        True  
4      True           House Ads          1/1/18           NaN        True

In [74]:
full_df = pd.merge(full_df, users_df, on='user_id', how='outer')

full_df.head()

user_id date_served marketing_channel          variant language_displayed  \
0   a1000      1/1/18         House Ads  personalization            English   
1   a1001      1/1/18         House Ads  personalization            English   
2   a1002      1/1/18         House Ads  personalization            English   
3   a1003      1/1/18         House Ads  personalization            English   
4   a1004      1/1/18         House Ads  personalization            English   

  converted subscribing_channel date_subscribed date_canceled is_retained  \
0      True           House Ads          1/1/18           NaN        True   
1      True           House Ads          1/1/18           NaN        True   
2      True           House Ads          1/1/18           NaN        True   
3      True           House Ads          1/1/18           NaN        True   
4      True           House Ads          1/1/18           NaN        True   

     age_group language_preferred  
0   0-18 years            English  
1  19-24 years            English  
2  24-30 years            English  
3  30-36 years            English  
4  36-45 years            English

In [75]:
full_df.sample(5)

user_id date_served marketing_channel          variant  \
9849  a10849     1/30/18         House Ads  personalization   
6677   a7677      1/9/18         Instagram  personalization   
5219   a6219     1/11/18         Instagram          control   
3962   a4962     1/11/18         House Ads          control   
6665   a7665      1/8/18          Facebook  personalization   

     language_displayed converted subscribing_channel date_subscribed  \
9849            English     False                 NaN             NaN   
6677            English     False                 NaN             NaN   
5219            English     False                 NaN             NaN   
3962            English     False                 NaN             NaN   
6665            English     False                 NaN             NaN   

     date_canceled is_retained    age_group language_preferred  
9849           NaN         NaN  19-24 years            English  
6677           NaN         NaN  30-36 years            English  
5219           NaN         NaN  30-36 years            Spanish  
3962           NaN         NaN  24-30 years            English  
6665           NaN         NaN  36-45 years            English

In [76]:
full_df.shape

(10037, 12)

In [77]:
#все типы object
full_df.dtypes

user_id                object
date_served            object
marketing_channel      object
variant                object
language_displayed     object
converted              object
subscribing_channel    object
date_subscribed        object
date_canceled          object
is_retained            object
age_group              object
language_preferred     object
dtype: object

In [78]:
full_df.describe().T

count unique          top  freq
user_id              10037  10037        a1000     1
date_served          10021     31      1/15/18   789
marketing_channel    10022      5    House Ads  4733
variant              10037      2      control  5091
language_displayed   10037      4      English  9793
converted            10022      2        False  8946
subscribing_channel   1856      5    Instagram   600
date_subscribed       1856     31      1/16/18   163
date_canceled          577    115       4/2/18    15
is_retained           1856      2         True  1279
age_group            10037      7  19-24 years  1682
language_preferred   10037      4      English  9275

In [79]:
full_df.isna().sum()

user_id                   0
date_served              16
marketing_channel        15
variant                   0
language_displayed        0
converted                15
subscribing_channel    8181
date_subscribed        8181
date_canceled          9460
is_retained            8181
age_group                 0
language_preferred        0
dtype: int64

Попробуем определить эффективность маркетинговых каналов по привлечению платящих игроков

Будем работать с полями:

marketing_channel: рекламный канал

converted: конверсия в покупку, 1 - пользователь совершил покупку, 0 - нет

In [80]:
#заменим пропущенные значение на моду
#full_df['marketing_channel'].fillna(full_df['marketing_channel'].mode())

In [81]:
#заменим пропущенные значение на моду
#full_df['converted'].fillna(full_df['converted'].mode())

In [82]:
full_df['converted'].value_counts()

False    8946
True     1076
Name: converted, dtype: int64

In [83]:
full_df[full_df['converted'] == True]['marketing_channel']

0        House Ads
1        House Ads
2        House Ads
3        House Ads
4        House Ads
           ...    
10030        Email
10031        Email
10032        Email
10033        Email
10035        Email
Name: marketing_channel, Length: 1076, dtype: object

In [96]:
full_df[full_df['converted'] == True].groupby('marketing_channel').count()['user_id']

marketing_channel
Email        193
Facebook     237
House Ads    298
Instagram    265
Push          83
Name: user_id, dtype: int64

In [89]:
full_df.groupby(['marketing_channel', 'converted']).count()['user_id']

marketing_channel  converted
Email              False         372
                   True          193
Facebook           False        1623
                   True          237
House Ads          False        4435
                   True          298
Instagram          False        1606
                   True          265
Push               False         910
                   True           83
Name: user_id, dtype: int64

In [91]:
full_df.groupby(['marketing_channel', 'converted']).count()['user_id'].unstack()

converted          False  True
marketing_channel             
Email                372   193
Facebook            1623   237
House Ads           4435   298
Instagram           1606   265
Push                 910    83

In [90]:
full_df.pivot_table(
    index='marketing_channel',
    columns='converted',
    values='user_id',
    aggfunc='count'
)

converted          False  True
marketing_channel             
Email                372   193
Facebook            1623   237
House Ads           4435   298
Instagram           1606   265
Push                 910    83

In [92]:
#теперь попробуем найти процентное соотношение (группировка с нормализацией)
pd.crosstab(index=full_df['marketing_channel'],
            columns=full_df['converted'],
            normalize='index')

converted             False      True
marketing_channel                    
Email              0.658407  0.341593
Facebook           0.872581  0.127419
House Ads          0.937038  0.062962
Instagram          0.858365  0.141635
Push               0.916415  0.083585

**Вывод**:

Наиболее эффективный канал в абсолютном выражении: 

House Ads	- 298 покупателей

В процентном соотношении конверсии:

Email	-	0.341593 конверсия

Попробуем определить количество игроков в каждой возрастной группе

In [98]:
full_df.groupby('age_group').count()['user_id']

age_group
0-18 years     1539
19-24 years    1682
24-30 years    1568
30-36 years    1355
36-45 years    1353
45-55 years    1353
55+ years      1187
Name: user_id, dtype: int64

In [99]:
#среди тех, кто конвертировался в покупку
full_df[full_df['converted'] == True].groupby('age_group').count()['user_id']

age_group
0-18 years     201
19-24 years    314
24-30 years    238
30-36 years     82
36-45 years     78
45-55 years     80
55+ years       83
Name: user_id, dtype: int64

**Вывод:**

В обоих случаях лидирует группа - 19-24 years





Попробуем определить самую раннюю дату подписки на сервис (поле "date_subscribed")

In [100]:
#преобразуем в datetime

from datetime import datetime, timedelta

full_df['date_subscribed'] = pd.to_datetime(full_df['date_subscribed'])

full_df.head(20)

user_id date_served marketing_channel          variant language_displayed  \
0    a1000      1/1/18         House Ads  personalization            English   
1    a1001      1/1/18         House Ads  personalization            English   
2    a1002      1/1/18         House Ads  personalization            English   
3    a1003      1/1/18         House Ads  personalization            English   
4    a1004      1/1/18         House Ads  personalization            English   
5    a1005      1/1/18         House Ads  personalization             German   
6    a1006      1/1/18         House Ads  personalization            English   
7    a1007      1/1/18         House Ads  personalization            English   
8    a1008      1/1/18         House Ads  personalization            English   
9    a1009      1/1/18         House Ads  personalization            English   
10   a1010      1/1/18         House Ads  personalization            English   
11   a1011      1/1/18         House Ads  personalization            English   
12   a1012      1/1/18         House Ads  personalization             Arabic   
13   a1013      1/1/18         House Ads  personalization            English   
14   a1014      1/1/18         House Ads  personalization            English   
15   a1015      1/1/18         House Ads  personalization             Arabic   
16   a1016      1/2/18         House Ads  personalization            English   
17   a1017      1/2/18         House Ads  personalization            English   
18   a1018      1/2/18         House Ads  personalization            English   
19   a1019      1/2/18         House Ads  personalization             German   

   converted subscribing_channel date_subscribed date_canceled is_retained  \
0       True           House Ads      2018-01-01           NaN        True   
1       True           House Ads      2018-01-01           NaN        True   
2       True           House Ads      2018-01-01           NaN        True   
3       True           House Ads      2018-01-01           NaN        True   
4       True           House Ads      2018-01-01           NaN        True   
5       True           House Ads      2018-01-01           NaN        True   
6       True           House Ads      2018-01-01           NaN        True   
7       True           House Ads      2018-01-01           NaN        True   
8       True           House Ads      2018-01-01           NaN        True   
9       True           House Ads      2018-01-01           NaN        True   
10      True           House Ads      2018-01-01           NaN        True   
11      True           House Ads      2018-01-01       1/18/18       False   
12      True           House Ads      2018-01-01       2/22/18       False   
13      True           House Ads      2018-01-01        3/9/18       False   
14      True           House Ads      2018-01-01       2/18/18       False   
15      True           House Ads      2018-01-01        3/3/18       False   
16      True           House Ads      2018-01-02        2/4/18       False   
17      True           House Ads      2018-01-02        1/9/18       False   
18      True           House Ads      2018-01-02        2/4/18       False   
19      True           House Ads      2018-01-02       1/21/18       False   

      age_group language_preferred  
0    0-18 years            English  
1   19-24 years            English  
2   24-30 years            English  
3   30-36 years            English  
4   36-45 years            English  
5   45-55 years             German  
6     55+ years            English  
7    0-18 years            English  
8   19-24 years            English  
9   24-30 years            English  
10   0-18 years            English  
11  19-24 years            English  
12  24-30 years             Arabic  
13   0-18 years            English  
14  19-24 years            English  
15  24-30 years             Arabic  
16   0-18 years            English  
17  19-24 years  

In [105]:
min_date = full_df['date_subscribed'].min()

print(f'Cамая ранняя дата подписки на сервис {min_date}')

Cамая ранняя дата подписки на сервис 2018-01-01 00:00:00


Попробуем определить портрет аудитории удержанных подписчиков (по возрасту и языку)

In [111]:
pd.pivot_table(data=full_df[full_df['is_retained'] == True],
               values='user_id',
               index='age_group',
               columns='language_preferred',
               aggfunc='count')

language_preferred  Arabic  English  German  Spanish
age_group                                           
0-18 years             3.0    210.0    10.0      6.0
19-24 years            1.0    304.0    11.0     10.0
24-30 years            3.0    250.0    15.0      5.0
30-36 years            NaN    114.0     1.0      7.0
36-45 years            1.0    107.0     2.0      5.0
45-55 years            2.0    108.0     3.0      6.0
55+ years              2.0     86.0     2.0      5.0